services/core/MasterDriverAgent/master_driver/interfaces/modbus_tk/client.py

```diff
@@ -0,0 +1,593 @@
+        elif self.table == REGISTER_READ_ONLY:
+            modbus_function_code = modbus_constants.READ_INPUT_REGISTERS
+        elif self.table == COIL_READ_WRITE:
+            modbus_function_code = modbus_constants.READ_COILS
+        elif self.table == COIL_READ_ONLY:
+            modbus_function_code = modbus_constants.READ_DISCRETE_INPUTS
+        else:
+            raise Exception("Unknown table type : {0}".format(self.table))
+
+        return modbus_function_code
+
+    @property
+    def write_function_code(self):
+        """Returns a modbus write function code appropriate for the table."""
+
+        if self.table == REGISTER_READ_WRITE:
+            # WRITE_MULTIPLE is used unless client cannot support it, see _write_single_values in Client.__init__
+            return modbus_constants.WRITE_MULTIPLE_REGISTERS
+        elif self.table == COIL_READ_WRITE:
+            # @TODO Temp hack around COIL write problem.
+            return modbus_constants.WRITE_SINGLE_COIL if self.count == 1 \
+                else modbus_constants.WRITE_MULTIPLE_COILS
+        else:
+            raise Exception("Unknown table type or write operation not supported : {0}".format(self.table))
+
+    @property
+    def formatting(self):
+        """
+            Provide a struct library format string for non-COIL data types.
+
+            For COILS, we leave the formatting to modbus-tk.
+        """
+        if self.table not in (COIL_READ_ONLY, COIL_READ_WRITE):
+            return self.data_format
+        else:
+            return None
+
+def compile_requests(fields, byte_order):
+    """
+
+    Creates a set of Modbus requests for the fields provided.  The fields
+    are sorted by table and address so that a minimum number of
+    requests can be created.
+
+    These requests are used for both reading and writing.
+
+    :param fields: List of fields sorted by address.
+    :param byte_order: Byte order of the modbus slave.
+    :return: List of Requests
+    """
+
+    register_count = 0
+    requests = list()
+
+    if not fields:
+        return requests
+
+    fields.sort(key=lambda f: f.table * 100000 + f.address)
+
+    # Set up the initial request before starting the loop
+    current_request = None
+    next_address = None
+
+    for f in fields:
+        # Decide if we need to start a new request
+        if ((current_request is None) or
+                (current_request.table != f.table or next_address != f.address) or
+                (f.type[LENGTH] > 1)):
+            current_request = Request(name=f.name, table=f.table, address=f.address, data_format=byte_order)
+            requests.append(current_request)
+            next_address = f.address
+
+        # Add the field to the current request.
+        # current_request.data_format += (f.type[FORMAT] * f.type[LENGTH])
+        try:
+            struct_format = str(f.type[LENGTH]) + f.type[FORMAT]
+        except TypeError:
+            print(f)
+            print(f)
```

Comment:
I think we should use more verbose logging messages here instead of print statements.

Reply:
Thanks. I fixed it.